### PR TITLE
Add authorization to Prometheus /metrics endpoint

### LINKED ISSUE #291: Add authorization to Prometheus scraping endpoint
Right now, if you go to `<server-url>/metrics` for any server, the Prometheus metrics will return even if you don't have any authentication token. We want to protect this endpoint by only allowing authenticated user to access it, although it might be a bit tricky to do so.

1. Issue a token signed by the server and attach it to the header of the Prometheus static scraper
2. Check token authentication at `/metrics` endpoint
3. Might need to come up a way to refresh the token periodically
4. Add regression tests

Note that this will break director's scraping mechanism. Will need a separate PR to allow director to carry a token when scraping origins' metrics.

### PR DESCRIPTION
**Pull request by haoming29**
Closes #291 

* Added Bearer authorization to Prometheus `/metrics` endpoint and issue token for the scraper of both director and server itself, with periodic reloading.
* Refactored authorization logic in `web_ui` to make it more flexible, clear, and reusable.
* Changed LoadDirectorPublicKey to return key, not pointer to the key

## Testing Instructions
* Spin up `director` and you should expect `/metrics` requests are all successful, and their frequency should be around 15s.
* Change `Monitoring.TokenExpiresIn` and `Monitoring.TokenRefreshInterval` to smaller value to ensure token refreshment is successful
* Run `director`, `registry` and `origin`. At `origin` end, confirm that its `/metrics` requests are also all successful.
* Go to origin's web APIs and check auth works as expected. i.e. `/api/v1.0/prometheus` as well as `/metrics` (You should be able to see it through web only when you logged in)

### CODE REVIEW

**Comment by haoming29** [comment]
@turetske Sorry I just found a new bug in this PR. Trying to address it so please hold on reviewing it. Thanks!

**Comment by haoming29** [comment]
> @turetske Sorry I just found a bug in this PR. Trying to address it so please hold on reviewing it. Thanks!

Fixed! Now it should be good to go.

File: origin_ui/origin.go [hunk 1]
```diff
@@ -184,9 +184,11 @@ func setLoginCookie(ctx *gin.Context, user string) {
 
 	issuerURL := url.URL{}
 	issuerURL.Scheme = "https"
-	issuerURL.Host = ctx.Request.URL.Host
+	issuerURL.Host = config.ComputeExternalAddress()
 	now := time.Now()
 	tok, err := jwt.NewBuilder().
+		// TODO: We might want to come up with some names broader than this for a
+		// generic token for Web APIs, like web_ui.access
 		Claim("scope", "prometheus.read").
```
Comment (haoming29) [inline origin_ui/origin.go:192]: This can be addressed in #308

**Review by turetske — APPROVED** [review]
LGTM!

**Pull request merged**